### PR TITLE
Implement `formatter` specializations for `pair` and `tuple`

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Towards #2919. Also partially implements LWG-3839 and LWG-3892.

Test coverage mainly relies on unblocked libcxx tests.
- `std/utilities/format/format.tuple/format.functions.format.pass.cpp`
- `std/utilities/format/format.tuple/format.functions.vformat.pass.cpp`
- `std/utilities/format/format.tuple/set_brackets.pass.cpp`
- `std/utilities/format/format.tuple/set_separator.pass.cpp`

There's one new test file added to the `std` subdirectory that is derived from libcxx tests.

The following ones would also be unblocked after this PR and LLVM-83734.
- `std/containers/sequences/vector.bool/vector.bool.fmt/parse.pass.cpp`
- `std/thread/thread.threads/thread.thread.class/thread.thread.id/parse.pass.cpp`
- `std/utilities/format/format.tuple/parse.pass.cpp`

### CODE REVIEW

**Comment by cpplearner** [comment]
We should also have test coverage in `std`. This helps to ensure that the feature works with different configurations (e.g. intelliSense).

File: stl/inc/format [hunk 1]
```diff
@@ -3934,6 +3938,251 @@ _NODISCARD size_t formatted_size(const locale& _Loc, const wformat_string<_Types
...
+    _Tuple_formatter_base()                                       = delete;
+    _Tuple_formatter_base(const _Tuple_formatter_base&)           = delete;
+    _Tuple_formatter_base operator=(const _Tuple_formatter_base&) = delete;
+};
+
+template <class _CharT, formattable<_CharT> _Ty1, formattable<_CharT> _Ty2>
+struct _Tuple_formatter_base<pair<_Ty1, _Ty2>, _CharT> : _Tuple_formatter_common_base<_CharT, _Ty1, _Ty2> {
+private:
+    using _Formatted_type =
+        _Maybe_const<_Tuple_formatter_common_base<_CharT, _Ty1, _Ty2>::_Is_const_formattable, pair<_Ty1, _Ty2>>;
```
Comment (StephanTLavavej) [inline stl/inc/format:4158]: Style: If these are going to wrap, we may as well introduce `_Base` typedefs to make this more readable.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

File: stl/inc/format [hunk 2]
```diff
@@ -3934,6 +3938,251 @@ _NODISCARD size_t formatted_size(const locale& _Loc, const wformat_string<_Types
...
+                if constexpr (_Idx != 0) {
+                    _STD _Copy_unchecked(_Separator._Unchecked_begin(), _Separator._Unchecked_end(), _Tmp_ctx.out());
+                }
+                _STD get<_Idx>(_Underlying).format(_Arg, _Tmp_ctx);
+            };
+            (_Single_writer.template operator()<_Indice>(_Args), ...);
+        }(index_sequence_for<_ArgTypes...>{});
+        _STD _Copy_unchecked(_Closing_bracket._Unchecked_begin(), _Closing_bracket._Unchecked_end(), _Tmp_ctx.out());
+
+        return _STD _Write_aligned(_Fmt_ctx.out(), static_cast<int>(_Tmp_buf.size()), _Format_specs, _Fmt_align::_Left,
```
[the remark this answers is not in the document]
Comment (cpplearner) [reply stl/inc/format:4107]:
```suggestion
        const auto _Width = (_STD min)(_Tmp_buf.size(), static_cast<size_t>((numeric_limits<int>::max)()));
        return _STD _Write_aligned(_Fmt_ctx.out(), static_cast<int>(_Width), _Format_specs, _Fmt_align::_Left,
```
(Should be done in a followup PR, of course.)

**Comment by StephanTLavavej** [comment]
`("Thank you!", "😻")`

**Pull request merged**